java-checks/src/main/java/org/sonar/java/checks/spring/SpelExpressionCheck.java

```diff
@@ -0,0 +1,246 @@
+      Stream.of(cls.modifiers().annotations()),
+      cls.members().stream().map(SpelExpressionCheck::getMemberAnnotations)
+    ).flatMap(Collection::stream);
+  }
+
+  private static List<AnnotationTree> getMemberAnnotations(Tree member) {
+    if (member.is(Tree.Kind.METHOD)) {
+      return ((MethodTree) member).modifiers().annotations();
+    } else if (member.is(Tree.Kind.VARIABLE)) {
+      return ((VariableTree) member).modifiers().annotations();
+    } else {
+      return Collections.emptyList();
+    }
+  }
+
+  private static boolean isSpringAnnotation(AnnotationTree annotation) {
+    return annotation.symbolType().fullyQualifiedName().startsWith(SPRING_PREFIX);
+  }
+
+  private void checkSpringAnnotationArguments(AnnotationTree annotation) {
+    annotation.arguments().stream()
+      .filter(argument -> argument.is(Tree.Kind.STRING_LITERAL))
+      .forEach(argument -> checkSpringAnnotationStringArgument((LiteralTree) argument));
+  }
+
+  private void checkSpringAnnotationStringArgument(LiteralTree argument) {
+    try {
+      var argValue = argument.value();
+      var argContents = argValue.substring(1, argValue.length() - 1);
+      parseStringContents(argContents);
+    } catch (SyntaxError e) {
+      reportIssue(argument, e);
+    }
+  }
+
+  private void reportIssue(LiteralTree stringToken, SyntaxError error) {
+    var tokenStart = Position.startOf(stringToken);
+    var textSpan = new AnalyzerMessage.TextSpan(
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeStart,
+      tokenStart.line(),
+      tokenStart.columnOffset() + error.rangeEnd
+    );
+
+    var analyzerMessage = new AnalyzerMessage(this, context.getInputFile(), textSpan, error.getMessage(), 0);
+    ((DefaultJavaFileScannerContext) context).reportIssue(analyzerMessage);
+  }
+
+  private static void parseStringContents(String content) {
+    var i = 0;
+    while (i < content.length()) {
+      var c = content.charAt(i);
+      switch (c) {
+        case '$':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidPropertyPlaceholder);
+          break;
+        case '#':
+          i = parseDelimitersAndContents(content, i + 1, i + 1, SpelExpressionCheck::checkValidSpelExpression);
+          break;
+        default:
+          i++;
+          break;
+      }
+    }
+  }
+
+  private static int parseDelimitersAndContents(
+    String value,
+    int startIndex,
+    int rangeStart,
+    ObjIntConsumer<String> parseContents) {
+    if (startIndex == value.length()) {
+      return startIndex;
+    }
+    var endIndex = parseDelimiterBraces(value, startIndex, rangeStart);
+    if (endIndex == startIndex) {
+      return endIndex;
+    }
+    var contents = value.substring(startIndex + 1, endIndex - 1);
+    parseContents.accept(contents, rangeStart);
+    return endIndex;
+  }
+
+  private static int parseDelimiterBraces(String value, int startIndex, int rangeStart) throws SyntaxError {
```

Comment:
I think the current name fits better, bacause:
1. it consomues the input (because it returns next char input) and also can throw a SyntaxError. Hence I think `parse` fits better here.
2. it also checks for the starting brace, not just the closing one, hence `DelimiterBraces`